File: leetecode/sortLinkedList.py
# Definition for singly-linked list.
from ast import main


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
"""Using merge sort"""
class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if(not head or not head.next):
            return head
        left = head
        right = self.getmid(left)
        temp = right.next
        right.next = None
        right = temp
        left = self.sortList(left)
        right = self.sortList(right)
        return self.merge(left,right)
    def merge(self,left,right):
        temp = tail = ListNode()
        while left and right:
            if(left.val<right.val):
                tail.next = left
                left = left.next
            else:
                tail.next = right
                right = right.next
            tail = tail.next
        if(left):
            tail.next = left
        if(right):
            tail.next = right
        return temp.next
    def getmid(self,head):
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

def main():
    head = ListNode(4)
    head.next = ListNode(2)
    head.next.next = ListNode(1)
    head.next.next.next = ListNode(3)
    head = Solution().sortList(head)
    while head:
        print(head.val)
        head = head.next
        
main()

File: leetecode/SortLinkedList.java

class SortedLinkedList {
    ListNode merge(ListNode head1,ListNode head2){
        if(head1==null) return head2;
        if(head2==null) return head1;
        ListNode dh = new ListNode(0);
        ListNode temp = dh;
        while (head1!=null && head2!=null) {
            if(head1.val<head2.val) {
                temp.next = head1;
                head1 = head1.next;
            }
            else{
                temp.next = head2;
                head2 = head2.next;
            }
            temp = temp.next;
        }
        if(head1==null) temp.next = head2;
            if(head2==null) temp.next = head1;
            
        return dh.next;
    }
    ListNode mergeSort(ListNode head){
        if(head==null || head.next==null) return head;
        ListNode mid = getMid(head);
        ListNode right = mid.next;
        mid.next = null;
        ListNode hleft = mergeSort(head);
        ListNode hright = mergeSort(right);
        return merge(hleft, hright);
    }
    public ListNode sortList(ListNode head) {
        return mergeSort(head);
    }
    ListNode getMid(ListNode head){
        if(head==null) return null;
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast!=null && fast.next!=null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
    public static void main(String[] args) {
        SortedLinkedList obj = new SortedLinkedList();
        ListNode root = new ListNode(10);
        root.next = new ListNode(4);
        root.next.next = new ListNode(3);
        root.next.next.next = new ListNode(4);
        root.next.next.next.next = new ListNode(20);
        ListNode pt = obj.sortList(root);
        while (pt!=null) {
            System.out.print(pt.val+" ");
            pt = pt.next;
        }
    }
}